reconcile/vault_replication.py

```diff
@@ -0,0 +1,258 @@
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultReplicationV1,
+    VaultInstanceV1,
+)
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPolicyV1,
+)
+
+from reconcile.utils import gql
+
+from typing import List, cast, Optional
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault,
+    dest_vault,
+    current_dest_version,
+    current_source_version,
+    path,
+):
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+        copy_secret(
+            dry_run=dry_run,
+            source_vault=source_vault,
+            dest_vault=dest_vault,
+            path=path,
+            secret_dict=secret_dict,
+        )
+
+
+def copy_secret(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    path: str,
+    secret_dict: dict,
+) -> None:
+    secret, src_version = source_vault.read_all_with_version(secret_dict)
+    write_dict = {"path": path, "data": secret}
+    logging.info(["replicate_vault_secret", src_version, path])
+    if not dry_run:
+        dest_vault.write(write_dict)
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: _VaultClient, dest_vault: _VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    _, version = source_vault.read_all_with_version(secret_dict)
+
+    try:
+        _, dest_version = dest_vault.read_all_with_version(secret_dict)
+        if dest_version is None and version is None:
+            # v1 secrets don't have version
+            copy_secret(
+                dry_run=dry_run,
+                source_vault=source_vault,
+                dest_vault=dest_vault,
+                path=path,
+                secret_dict=secret_dict,
+            )
+        elif dest_version < version:
+            deep_copy_versions(
+                dry_run=dry_run,
+                source_vault=source_vault,
+                dest_vault=dest_vault,
+                current_dest_version=dest_version,
+                current_source_version=version,
+                path=path,
+            )
+        else:
+            logging.info(["replicate_vault_secret", dest_version, version, path])
+    except SecretNotFound:
+        logging.info(["replicate_vault_secret", "Secret not found", path])
+        deep_copy_versions(
+            dry_run=dry_run,
+            source_vault=source_vault,
+            dest_vault=dest_vault,
+            current_dest_version=0,
+            current_source_version=version,
+            path=path,
+        )
+
+
+def check_invalid_paths(
+    path_list: List[str],
+    policy_paths: Optional[List[str]],
+) -> None:
+
+    invalid_paths = []
+    if policy_paths is not None:
+        invalid_paths = list_invalid_paths(path_list, policy_paths)
+    if invalid_paths:
```

Comment:
You can put this inside the `if policy_paths is not None` condition

Reply:
👍